File: python/bob/bob.py
import re


def response(hey_bob):
    hey_bob = hey_bob.rstrip()
    if not hey_bob:
        return "Fine. Be that way!"

    is_shout = hey_bob.isupper()
    is_question = hey_bob.endswith('?')

    if is_shout and is_question:
        return "Calm down, I know what I'm doing!"
    if is_shout:
        return "Whoa, chill out!"
    if is_question:
        return "Sure."

    return "Whatever."

File: kotlin/bob/src/main/kotlin/Bob.kt
object Bob {
    fun hey(input: String): String {
        return when {
            input.isEmpty() -> "Fine. Be that way!"
            input.isQuestion() && input.isYelling() -> "Calm down, I know what I'm doing!"
            input.isQuestion() -> "Sure."
            input.isYelling() -> "Whoa, chill out!"
            input.isSilence() -> "Fine. Be that way!"
            else -> "Whatever."
        }
    }

    private fun String.isQuestion(): Boolean = if (this.trimEnd().isEmpty()) false else this.trimEnd().last() == '?'
    private fun String.isYelling(): Boolean = this.filter { it.isLetter() }.all { it.isUpperCase() }
            && this.any { it.isUpperCase() }
    private fun String.isSilence(): Boolean = this.isEmpty() || this.all { it.isWhitespace() }
}
